go/pkg/rexec/rexec.go

```diff
@@ -332,35 +341,31 @@ func (ec *Context) ExecuteRemotely() {
 	ec.Metadata.RealBytesUploaded = bytesMoved
 	log.V(1).Infof("%s %s> Executing remotely...\n%s", cmdID, executionID, strings.Join(ec.cmd.Args, " "))
 	ec.Metadata.EventTimes[command.EventExecuteRemotely] = &command.TimeInterval{From: time.Now()}
-	eg, ctx := errgroup.WithContext(ec.ctx)
-	var streamOut, streamErr sync.Once
-	op, err := ec.client.GrpcClient.ExecuteAndWaitProgress(ctx, &repb.ExecuteRequest{
+	// These variables are owned by the progress callback (which is async but not conrrent) until the execution returns.
+	var streaming bool
+	var nOutStreamed, nErrStreamed int64
+	op, err := ec.client.GrpcClient.ExecuteAndWaitProgress(ec.ctx, &repb.ExecuteRequest{
 		InstanceName:    ec.client.GrpcClient.InstanceName,
 		SkipCacheLookup: !ec.opt.AcceptCached || ec.opt.DoNotCache,
 		ActionDigest:    ec.Metadata.ActionDigest.ToProto(),
 	}, func(md *repb.ExecuteOperationMetadata) {
-		if !ec.opt.StreamOutErr {
+		if streaming || !ec.opt.StreamOutErr {
 			return
 		}
+		streaming = true
 		if name := md.GetStdoutStreamName(); name != "" {
-			streamOut.Do(func() {
-				eg.Go(func() error {
-					path := fmt.Sprintf("%s/logstreams/%s", ec.client.GrpcClient.InstanceName, name)
-					log.V(1).Infof("%s %s> Streaming to stdout from %q", cmdID, executionID, path)
-					_, err = ec.client.GrpcClient.ReadResourceTo(ctx, path, outerr.NewOutWriter(ec.oe))
-					return err
-				})
-			})
+			path := fmt.Sprintf("%s/logstreams/%s", ec.client.GrpcClient.InstanceName, name)
+			log.V(1).Infof("%s %s> Streaming to stdout from %q", cmdID, executionID, path)
+			// Ignoring the error here since the net result is downloading the full stream after the fact.
+			n, _ := ec.client.GrpcClient.ReadResourceTo(ec.ctx, path, outerr.NewOutWriter(ec.oe))
+			nOutStreamed += n
 		}
 		if name := md.GetStderrStreamName(); name != "" {
-			streamErr.Do(func() {
-				eg.Go(func() error {
-					path := fmt.Sprintf("%s/logstreams/%s", ec.client.GrpcClient.InstanceName, name)
-					log.V(1).Infof("%s %s> Streaming to stdout from %q", cmdID, executionID, path)
-					_, err = ec.client.GrpcClient.ReadResourceTo(ctx, path, outerr.NewErrWriter(ec.oe))
-					return err
-				})
-			})
+			path := fmt.Sprintf("%s/logstreams/%s", ec.client.GrpcClient.InstanceName, name)
+			log.V(1).Infof("%s %s> Streaming to stdout from %q", cmdID, executionID, path)
+			// Ignoring the error here since the net result is downloading the full stream after the fact.
+			n, _ := ec.client.GrpcClient.ReadResourceTo(ec.ctx, path, outerr.NewErrWriter(ec.oe))
```

Comment:
I think the net effect here is that we won't even try to download stderr until the action is finished, since we won't get here until stdout is finished being read from above.

Reply:
I'm not following you here. The download below always happens after the execution call has returned.

Reply:
Sorry for the ambiguous wording. :) I was referring to the streaming download that happens on this line (not the blob download that takes place after the action has completed if necessary). Because you had removed the goroutines and called ReadResourceTo for both stdout and stderr, one after the other, and synchronously, the first ReadResourceTo call (for stdout) wouldn't finish until the action completes (because the API server won't close the stdout stream until the action completes), and only then would we start "streaming" (not really streaming) stderr. The intent was to be streaming both concurrently while the action is still running.

Now that you've added goroutines back, this comment is resolved.